TRABAJO PRÁCTICO: HERRAMIENTAS PARA EL ANÁLISIS DE DATOS

# Importación de librerías a utilizar

In [37]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

#Importación de Dataset

In [38]:
data_set_pc=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/precios-en-surtidor-resolucin-3142016.csv')

In [39]:
data_set_pc

,indice_tiempo,idempresa,cuit,empresa,direccion,localidad,provincia,region,idproducto,producto,idtipohorario,tipohorario,precio,fecha_vigencia,idempresabandera,empresabandera,latitud,longitud,geojson
0,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,2,Diurno,2190.0,2026-05-01 12:43:00,28,PUMA,-34.658476,-58.529443,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."
1,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,3,Nocturno,2190.0,2026-05-01 12:43:00,28,PUMA,-34.658476,-58.529443,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."
2,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,21,Gas Oil Grado 3,2,Diurno,2478.0,2026-05-01 12:44:00,28,PUMA,-34.658476,-58.529443,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."
3,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,21,Gas Oil Grado 3,3,Nocturno,2479.0,2026-05-01 12:44:00,28,PUMA,-34.658476,-58.529443,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."
4,2026-04,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,6,GNC,2,Diurno,599.9,2026-04-02 14:09:00,28,PUMA,-34.658476,-58.529443,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
36754,2026-05,4097,30-70795820-7,ZUSICO S.A.,TALCAHUANO 1198,MAR DEL PLATA,BUENOS AIRES,PAMPEANA,6,GNC,3,Nocturno,1049.0,2026-05-18 10:00:00,28,PUMA,-38.031169,-57.575518,"{""type"":""Point"",""coordinates"":[-57.575518,-38...."
36755,2025-08,4097,30-70795820-7,ZUSICO S.A.,TALCAHUANO 1198,MAR DEL PLATA,BUENOS AIRES,PAMPEANA,3,Nafta (premium) de más de 95 Ron,2,Diurno,1757.0,2025-08-02 09:00:00,28,PUMA,-38.031169,-57.575518,"{""type"":""Point"",""coordinates"":[-57.575518,-38...."
36756,2025-08,4097,30-70795820-7,ZUSICO S.A.,TALCAHUANO 1198,MAR DEL PLATA,BUENOS AIRES,PAMPEANA,3,Nafta (premium) de más de 95 Ron,3,Nocturno,1757.0,2025-08-02 09:00:00,28,PUMA,-38.031169,-57.575518,"{""type"":""Point"",""coordinates"":[-57.575518,-38...."
36757,2026-05,4097,30-70795820-7,ZUSICO S.A.,TALCAHUANO 1198,MAR DEL PLATA,BUENOS AIRES,PAMPEANA,2,Nafta (súper) entre 92 y 95 Ron,2,Diurno,2178.0,2026-05-15 09:00:00,28,PUMA,-38.031169,-57.575518,"{""type"":""Point"",""coordinates"":[-57.575518,-38...."


## Visualización general: cantidad de columnas y filas

In [40]:
data_set_pc.shape

(36759, 19)

## Información general del Dataset

In [41]:
data_set_pc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36759 entries, 0 to 36758
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   indice_tiempo     36759 non-null  object 
 1   idempresa         36759 non-null  int64  
 2   cuit              36759 non-null  object 
 3   empresa           36759 non-null  object 
 4   direccion         36759 non-null  object 
 5   localidad         36759 non-null  object 
 6   provincia         36759 non-null  object 
 7   region            28788 non-null  object 
 8   idproducto        36759 non-null  int64  
 9   producto          36759 non-null  object 
 10  idtipohorario     36759 non-null  int64  
 11  tipohorario       36759 non-null  object 
 12  precio            36759 non-null  float64
 13  fecha_vigencia    36759 non-null  object 
 14  idempresabandera  36759 non-null  int64  
 15  empresabandera    36759 non-null  object 
 16  latitud           36733 non-null  float6

### Tipos de datos del Dataset

In [42]:
data_set_pc.dtypes

,0
indice_tiempo,object
idempresa,int64
cuit,object
empresa,object
direccion,object
localidad,object
provincia,object
region,object
idproducto,int64
producto,object


### Descripción general del Dataset

In [43]:
data_set_pc.describe()

,idempresa,idproducto,idtipohorario,precio,idempresabandera,latitud,longitud
count,36759.000000,36759.000000,36759.000000,36759.000000,36759.000000,36733.000000,36733.000000
mean,5553.796104,10.732963,2.500014,1460.834187,8.944068,-33.416583,-61.603435
std,3413.933124,8.477581,0.500007,602.430246,10.775520,4.108302,3.622760
min,104.000000,2.000000,2.000000,6.999000,-1.000000,-54.829501,-72.338930
25%,2333.000000,3.000000,2.000000,1245.000000,2.000000,-34.804110,-64.268935
50%,5748.000000,6.000000,3.000000,1435.000000,4.000000,-34.062579,-60.672217
75%,8471.000000,19.000000,3.000000,1760.000000,26.000000,-31.444160,-58.552110
max,11501.000000,21.000000,3.000000,3000.000000,30.000000,-22.055820,-53.648260


Al analizar la madia y la mediana, podemos observar que, en general, el grupo de datos es homogeneo ya que no hay diferencias significativas entre estas dos variables


## OUTLIERS

In [44]:
def find_outliers_iqr(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    return outliers

In [45]:
# Identificar outliers en la columna 'precio' usando la función
outliers_precio = find_outliers_iqr(data_set_pc, 'precio')

print(f"Se encontraron {len(outliers_precio)} outliers en la columna 'precio'.")
print("Valores atípicos en 'precio' (primeros 5):")
display(outliers_precio.head())

Se encontraron 3565 outliers en la columna 'precio'.
Valores atípicos en 'precio' (primeros 5):


,indice_tiempo,idempresa,cuit,empresa,direccion,localidad,provincia,region,idproducto,producto,idtipohorario,tipohorario,precio,fecha_vigencia,idempresabandera,empresabandera,latitud,longitud,geojson
142,2026-04,2306,30-68604387-4,9 DE JULIO COMBUSTIBLES SRL,AV. 9 DE JULIO 1501,RESISTENCIA,CHACO,NEA,21,Gas Oil Grado 3,2,Diurno,2538.00,2026-04-20 14:00:00,26,AXION,-27.463010,-58.973050,"{""type"":""Point"",""coordinates"":[-58.97305,-27.4..."
143,2026-04,2306,30-68604387-4,9 DE JULIO COMBUSTIBLES SRL,AV. 9 DE JULIO 1501,RESISTENCIA,CHACO,NEA,21,Gas Oil Grado 3,3,Nocturno,2538.00,2026-04-20 14:00:00,26,AXION,-27.463010,-58.973050,"{""type"":""Point"",""coordinates"":[-58.97305,-27.4..."
168,2026-05,3728,30-50646337-4,ABDALA HNOS.S.R.L.,LIBERTAD 2080,SANTIAGO DEL ESTERO,SANTIAGO DEL ESTERO,NOA,21,Gas Oil Grado 3,2,Diurno,2604.00,2026-05-15 20:38:00,4,SHELL C.A.P.S.A.,-27.798330,-64.277230,"{""type"":""Point"",""coordinates"":[-64.27723,-27.7..."
169,2026-05,3728,30-50646337-4,ABDALA HNOS.S.R.L.,LIBERTAD 2080,SANTIAGO DEL ESTERO,SANTIAGO DEL ESTERO,NOA,21,Gas Oil Grado 3,3,Nocturno,2604.00,2026-05-15 20:38:00,4,SHELL C.A.P.S.A.,-27.798330,-64.277230,"{""type"":""Point"",""coordinates"":[-64.27723,-27.7..."
186,2017-12,348,30-60588371-7,ABELLA Y ALONSO S.R.L.,MORENO 1701,VILLA BALLESTER,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,2,Diurno,22.99,2017-12-06 11:34:00,1,BLANCA,-34.533377,-58.557665,"{""type"":""Point"",""coordinates"":[-58.557665,-34...."


# Preguntas para el análisis del dataset:
1) ¿Cuál es la variación del precio de los combustibles según el momento del día (mañana, tarde o noche) en que se realiza la carga?
2) ¿Cuál es la variación del precio de los combustibles según la empresa de bandera?
3) ¿Cuál es el precio promedio de cada tipo de combustible?

Lo primero que voy a hacer es eliminar las columnas que no voy a utilizar para el analsis. En este caso la latitud y longitud

In [46]:
data_set_pc = data_set_pc.drop(columns=['latitud', 'longitud'])

In [47]:
data_set_pc.head(5)

,indice_tiempo,idempresa,cuit,empresa,direccion,localidad,provincia,region,idproducto,producto,idtipohorario,tipohorario,precio,fecha_vigencia,idempresabandera,empresabandera,geojson
0,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,2,Diurno,2190.0,2026-05-01 12:43:00,28,PUMA,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."
1,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,3,Nocturno,2190.0,2026-05-01 12:43:00,28,PUMA,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."
2,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,21,Gas Oil Grado 3,2,Diurno,2478.0,2026-05-01 12:44:00,28,PUMA,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."
3,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,21,Gas Oil Grado 3,3,Nocturno,2479.0,2026-05-01 12:44:00,28,PUMA,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."
4,2026-04,1376,33-64337382-9,10 DE SETIEMBRE S.A.,Av. Mosconi 299,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,6,GNC,2,Diurno,599.9,2026-04-02 14:09:00,28,PUMA,"{""type"":""Point"",""coordinates"":[-58.529443,-34...."


In [48]:
data_set_pc = data_set_pc.drop(columns=['idempresabandera','geojson','fecha_vigencia','idtipohorario','direccion'])

In [49]:
data_set_pc.head()

,indice_tiempo,idempresa,cuit,empresa,localidad,provincia,region,idproducto,producto,tipohorario,precio,empresabandera
0,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,Diurno,2190.0,PUMA
1,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,Nocturno,2190.0,PUMA
2,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,21,Gas Oil Grado 3,Diurno,2478.0,PUMA
3,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,21,Gas Oil Grado 3,Nocturno,2479.0,PUMA
4,2026-04,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,6,GNC,Diurno,599.9,PUMA


Voy a contar cuantos surtidores se relevaron teniendo en cuenta la emrpesa de bandera

In [50]:
data_set_pc.duplicated().sum()

np.int64(32)

Entiendo que esto significa que no hay duplicados.

Voy a sacar valores nulos

In [51]:
data_set_pc.isnull().sum()

,0
indice_tiempo,0
idempresa,0
cuit,0
empresa,0
localidad,0
provincia,0
region,7971
idproducto,0
producto,0
tipohorario,0


Entiendo que en la columna "region" hay 7971 datos nulos, pero como no es una variable a analizar, no lo tengo en cuenta.

Voy a buscar datos duplicados

In [52]:
data_set_pc.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
36754,False
36755,False
36756,False
36757,False


No hay ningún dato duplicado

Voy a cambiarle el nombre a algunas columnas, para que sea más facil identificar a que hacen referencia los datos que contienen

In [53]:
data_set_pc = data_set_pc.rename(columns={"indice_tiempo": "fecha", "empresa":"razon social", "empresabandera":"nombre_comercial"})
print("Columns after renaming:", data_set_pc.columns.tolist())

Columns after renaming: ['fecha', 'idempresa', 'cuit', 'razon social', 'localidad', 'provincia', 'region', 'idproducto', 'producto', 'tipohorario', 'precio', 'nombre_comercial']


In [54]:
data_set_pc.head(5)

,fecha,idempresa,cuit,razon social,localidad,provincia,region,idproducto,producto,tipohorario,precio,nombre_comercial
0,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,Diurno,2190.0,PUMA
1,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,Nocturno,2190.0,PUMA
2,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,21,Gas Oil Grado 3,Diurno,2478.0,PUMA
3,2026-05,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,21,Gas Oil Grado 3,Nocturno,2479.0,PUMA
4,2026-04,1376,33-64337382-9,10 DE SETIEMBRE S.A.,LOMAS DEL MIRADOR,BUENOS AIRES,PAMPEANA,6,GNC,Diurno,599.9,PUMA


 El dataset: Precio de combustible por región

Cada fila es **una variedad de combustible**:

| Columna | Significado | Tipo de variable |
|---|---|---|
| `fecha` | Fecha de medición | Numérica |
| `idempresa` | Numero de ID de la empresa | Numérica |
| `CUIT` | Numero de CUIT | Numérica |
| `razon social` | Razon social | Categórica |
| `Producto` | Variedad de combustible | Categórica |
| `Tipohorario` | Momento del dia de carga | Categorica |
| `precio` | Valor del combustible | Numérica |
| `nombre_comercial` | Empresa a la que pertenece | Categórica |

Productos y su frecuencia

In [55]:
product_counts = data_set_pc['producto'].value_counts()
display(product_counts)

,count
producto,
Gas Oil Grado 2,8372
Nafta (súper) entre 92 y 95 Ron,8348
Gas Oil Grado 3,8194
Nafta (premium) de más de 95 Ron,8125
GNC,3720


Precios Promedio de los Productos
A continuación, calcularemos el precio promedio para cada tipo de combustible en el dataset.

In [56]:
average_prices = data_set_pc.groupby('producto')['precio'].mean().reset_index()
average_prices = average_prices.sort_values(by='precio', ascending=False)
display(average_prices)

,producto,precio
2,Gas Oil Grado 3,1678.583619
3,Nafta (premium) de más de 95 Ron,1644.660794
1,Gas Oil Grado 2,1473.804205
4,Nafta (súper) entre 92 y 95 Ron,1428.525276
0,GNC,623.011543


In [57]:
average_prices['precio'] = average_prices['precio'].round(2)
display(average_prices)

,producto,precio
2,Gas Oil Grado 3,1678.58
3,Nafta (premium) de más de 95 Ron,1644.66
1,Gas Oil Grado 2,1473.80
4,Nafta (súper) entre 92 y 95 Ron,1428.53
0,GNC,623.01


# Análisis gráfico

Gráfico de Barras de Precios Promedio por Producto
Este gráfico de barras muestra el precio promedio de cada tipo de combustible, lo que facilita la comparación visual entre ellos.


In [58]:
average_prices = data_set_pc.groupby('producto')['precio'].mean().reset_index()
average_prices = average_prices.sort_values(by='precio', ascending=False)
average_prices['precio'] = average_prices['precio'].round(2)

fig = px.bar(average_prices, x='producto', y='precio',
             title='Precio Promedio por Tipo de Combustible',
             labels={'producto': 'Tipo de Combustible', 'precio': 'Precio Promedio'})
fig.update_layout(xaxis_title_standoff=25) # Ajusta el espaciado del título del eje X
fig.show()

Por lo que puedo interpretar, el Gas Oil Grado 3 y la Nafta Premium tienen casi el mismo valor, asi como el Gas Oil Grado 2 y la Nafta Super. El GNC es el producto de menor valor, por lo que explica que los automoviles que se utilizan para trabajr (taxis, aplicaciones, etc) utilicen este combustible para abaratar sus costos.

Histograma Interactivo de Precios por Momento del Día (Plotly)
Este histograma interactivo, generado con Plotly, permite explorar la distribución de los precios de combustible (precio) para cada momento del día (tipohorario).

In [59]:
fig = px.histogram(data_set_pc, y='tipohorario', x='precio',
                   title='Distribución Interactiva de Precios de Combustible por Momento del Día',
                   labels={'precio': 'Precio del Combustible', 'tipohorario': 'Momento del Día de Carga'},
                   histnorm='density', # o 'count' si prefieres la cuenta de observaciones
                   nbins=50, # Número de "contenedores" en el histograma
                   marginal='box', # Añade un boxplot marginal para cada grupo
                   orientation='h') # Orientación horizontal para el histograma


In [60]:
fig.update_layout(bargap=0.1) # Espacio entre las barras
fig.show()

Por lo que interpreto, no varia el precio del combustible de manera significativa, segun el momento del día

Se puede ver que las marcas que tienen cajas mas grandes tienen mayor variación de precios, mientras que los que tienen cajas mas chicas sus precios varian poco, aunque por ejemplo YPF, posee muchos valores outliers, lo que indica que posee muchos valores por arriba y debajo de los tipicos.

In [61]:
fig_outliers = px.box(data_set_pc, y='precio',
                      title='Detección de Outliers en Precios de Combustible',
                      labels={'precio': 'Precio del Combustible'})
fig_outliers.show()

In [62]:
def find_outliers_iqr(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    return outliers

# Identificar outliers en la columna 'precio'
outliers_precio = find_outliers_iqr(data_set_pc, 'precio')

print(f"Se encontraron {len(outliers_precio)} outliers en la columna 'precio'.")
print("Valores atípicos en 'precio':")
display(outliers_precio.head())

Se encontraron 3565 outliers en la columna 'precio'.
Valores atípicos en 'precio':


,fecha,idempresa,cuit,razon social,localidad,provincia,region,idproducto,producto,tipohorario,precio,nombre_comercial
142,2026-04,2306,30-68604387-4,9 DE JULIO COMBUSTIBLES SRL,RESISTENCIA,CHACO,NEA,21,Gas Oil Grado 3,Diurno,2538.00,AXION
143,2026-04,2306,30-68604387-4,9 DE JULIO COMBUSTIBLES SRL,RESISTENCIA,CHACO,NEA,21,Gas Oil Grado 3,Nocturno,2538.00,AXION
168,2026-05,3728,30-50646337-4,ABDALA HNOS.S.R.L.,SANTIAGO DEL ESTERO,SANTIAGO DEL ESTERO,NOA,21,Gas Oil Grado 3,Diurno,2604.00,SHELL C.A.P.S.A.
169,2026-05,3728,30-50646337-4,ABDALA HNOS.S.R.L.,SANTIAGO DEL ESTERO,SANTIAGO DEL ESTERO,NOA,21,Gas Oil Grado 3,Nocturno,2604.00,SHELL C.A.P.S.A.
186,2017-12,348,30-60588371-7,ABELLA Y ALONSO S.R.L.,VILLA BALLESTER,BUENOS AIRES,PAMPEANA,19,Gas Oil Grado 2,Diurno,22.99,BLANCA


In [64]:
data_set_pc.to_csv('filename.csv', index=False)